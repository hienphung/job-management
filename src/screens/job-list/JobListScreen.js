import React, {Component} from "react";
import {Dimensions, Image, ScrollView, View} from "react-native";
import styles from "./styles";
import Calendar from "../../components/calendar/Calendar";
import {connect} from "react-redux";
import * as JobsActions from "../../redux/jobs/actions";
import {calendarData, imageSliderData, jobList} from "./fake-data";
import Carousel, {Pagination} from "react-native-snap-carousel";
import JobSection from "../../components/job-section/JobSection";

const {width: windowWidth} = Dimensions.get("window");

class JobListScreen extends Component {

	constructor(props) {
		super(props);

		this.state = {
			sliderCurrentIndex: 0
		};

		this.calendarOnItemPress = this.calendarOnItemPress.bind(this);
		this.carouselOnSnapToItem = this.carouselOnSnapToItem.bind(this);
		this.onJobItemPress = this.onJobItemPress.bind(this);
	}

	componentDidMount() {
		this.props.dispatch(JobsActions.setCalendarData(calendarData));
		this.props.dispatch(JobsActions.setCalendarSelectedIndex(3));
		this.props.dispatch(JobsActions.setJobListImages(imageSliderData));
		this.setState({sliderCurrentIndex: 1});
		this.props.dispatch(JobsActions.setJobList(jobList));
	}

	calendarOnItemPress(index) {
		this.props.dispatch(JobsActions.setCalendarSelectedIndex(index))
	}

	carouselOnSnapToItem(index) {
		this.setState({sliderCurrentIndex: index})
	}

	renderJobListImageItem({item, index}, parallaxProps) {
		return (
			<View style={styles.imageSliderItemContainer}>
				<Image
					key={item.id}
					resizeMode={"cover"}
					source={{uri: item.url}}
					containerStyle={styles.sliderImage}
					style={styles.sliderImage}
					parallaxFactor={1}
					{...parallaxProps}
				/>
			</View>
		)
	}

	renderImageSlider() {
		const {jobListImages} = this.props;
		if (jobListImages && jobListImages.length > 0) {
			const sliderWidth = windowWidth;
			const itemWidth = windowWidth;
			return (
				<View style={styles.imageSliderContainer}>
					<Carousel
						firstItem={1}
						ref={(c) => {
							this._carousel = c;
						}}
						data={jobListImages}
						renderItem={this.renderJobListImageItem}
						sliderWidth={sliderWidth}
						itemWidth={itemWidth}
						onSnapToItem={this.carouselOnSnapToItem}
					/>
					{jobListImages.length > 1 ? (
						<Pagination containerStyle={styles.sliderPagination}
						            activeDotIndex={this.state.sliderCurrentIndex}
						            dotsLength={jobListImages.length}
						            dotStyle={styles.sliderPaginationDot}
						            inactiveDotOpacity={0.4}
						            inactiveDotScale={0.6}/>
					) : null}
				</View>
			)
		}
		return null;
	}

	onJobItemPress(index, sectionIndex) {
		const selectedJob = this.props.jobList[sectionIndex].jobs[index];
		if (selectedJob) {
			this.props.navigation.navigate("JobDetails", {selectedJob})
		}
	}

	renderJobsSection(value, index) {
		return (
			<JobSection key={value.sectionId}
			            data={value.jobs}
			            containerStyle={styles.jobSection}
			            onItemPress={this.onJobItemPress}
									sectionIndex={index}/>
		)
	}

	renderJobsList() {
		const {jobList} = this.props;
		if (jobList && jobList.length > 0) {
			return (
				<View style={styles.jobSectionsContainer}>
					{jobList.map((value, index) => {
						return this.renderJobsSection(value, index);
					})}
				</View>
			)
		}
	}

	render() {
		const {calendarData, calendarSelectedIndex} = this.props;

		return (
			<ScrollView style={{flex: 1}} contentContainerStyle={styles.container}>
				<Calendar testId={"calendar"} data={calendarData}
				          style={styles.calendar}
				          selectedIndex={calendarSelectedIndex}
				          onItemPress={this.calendarOnItemPress}/>
				{this.renderImageSlider()}
				{this.renderJobsList()}
			</ScrollView>
		)
	}
}

const mapStateToProps = (state) => ({
	calendarData: state.jobs.calendarData,
	calendarSelectedIndex: state.jobs.calendarSelectedIndex,
	jobListImages: state.jobs.jobListImages,
	jobList: state.jobs.jobList
});

export default connect(mapStateToProps)(JobListScreen);