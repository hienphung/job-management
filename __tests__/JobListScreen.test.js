/**
 * @jest-environment jsdom
 */

import React from 'react';
import {shallow} from 'enzyme';
import JobListScreen from "../src/screens/job-list/JobListScreen.js";

describe("Test JobListScreen", () => {
	let screen;
	let calendarComponent;
	beforeAll(() => {
		screen = shallow(<JobListScreen/>);
		calendarComponent = screen.find("Calendar").first();
	});

	it("Should render correctly", () => {
		expect(screen).toMatchSnapshot();
	});
	it('Should render Calendar', () => {
		expect(calendarComponent).toBeTruthy();
	});
});