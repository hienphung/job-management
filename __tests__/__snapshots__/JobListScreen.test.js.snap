// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test JobListScreen Should render correctly 1`] = `ShallowWrapper {}`;
